substrate/client/network/sync/src/lib.rs

```diff
@@ -1248,30 +1184,18 @@ where
 	}
 
 	fn send_block_request(&mut self, who: PeerId, request: BlockRequest<B>) {
-		let (tx, rx) = oneshot::channel();
-		let opaque_req = self.create_opaque_block_request(&request);
-
 		if self.peers.contains_key(&who) {
-			self.pending_responses
-				.insert(who, Box::pin(async move { (who, PeerRequest::Block(request), rx.await) }));
-		}
-
-		match self.encode_block_request(&opaque_req) {
-			Ok(data) => {
-				self.network_service.start_request(
-					who,
-					self.block_request_protocol_name.clone(),
-					data,
-					tx,
-					IfDisconnected::ImmediateError,
-				);
-			},
-			Err(err) => {
-				log::warn!(
-					target: LOG_TARGET,
-					"Failed to encode block request {opaque_req:?}: {err:?}",
-				);
-			},
+			let downloader = self.block_downloader.clone();
+			self.pending_responses.insert(
+				who,
+				Box::pin(async move {
+					(
+						who,
+						PeerRequest::Block(request.clone()),
+						downloader.download_block(who, request).await,
+					)
+				}),
+			);
```

Comment:
This is something I would like to get refactored in the future. It's not related to this PR but you should know since you're working on your own syncing implementation.

I'd prefer having something like this:

```rust
enum RequestResponseError {
    Timeout
    Rejected
}

enum RequestResponseEvent {
    RequestReceived {
        peer: PeerId,
        request: Vec<u8>
    },
    ResponseReceived {
        peer: PeerId,
        response: Vec<u8>
    },
    RequestFailed {
        peer: PeerId,
        error: RequestResponseError,
    }
}

impl RequestResponseHandle {
    async fn send_request(peer: PeerId, request: Vec<u8>) {}
    async fn send_response(peer: PeerId, response: Vec<u8>) {}
    fn reject_request(peer: PeerId) {}
    async fn next_event(&mut self) -> Option<RequestResponseEvent> {}
}
```

Each reqeuest/response protocol would have its own handle which would allow sending requests/responses and would also allow polling both using the same object. This would eliminate using `FuturesUnordered` for polling responses. This would mean that both request and response processing is moved to `SyncingEngine`.

Reply:
Thanks for the insights, I'll try to implement this after requests-responses are moved to `SyncingEngine`.

Reply:
This is not as critical as the rest of the Sync 2.0 refactoring and it's more a future nice to have task, mainly for providing a better abstraction over libp2p.

Reply:
Thanks for the inputs, resolving as this is won't be part of this PR